File: 78-subsets.py
# create combinations of length 1 - len(nums)
# Time: ~O(n * 2^n)
# Space: ~O(n) - only uses space for output
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        def backtrack(start, combo):
            if len(combo) > len(nums):
                return

            combos.append(combo[:])

            for i in range(start, len(nums)):
                combo.append(nums[i])
                backtrack(i + 1, combo)
                combo.pop()

        combos = []
        backtrack(0, [])
        return combos

File: 78-subsets.cs
public class Solution {
    public IList<IList<int>> Subsets(int[] nums)
    {
        List<IList<int>> combos = new();
        dfs(0, new());
        return combos;

        void dfs(int start, List<int> combo)
        {
            if (combo.Count > nums.Length)
            {
                return;
            }

            combos.Add(new List<int>(combo));

            for (int i = start; i < nums.Length; i++)
            {
                combo.Add(nums[i]);
                dfs(i + 1, combo);
                combo.RemoveAt(combo.Count() - 1);
            }
        }
    }
}
